File: src/main/java/at/meks/calculation/Battery.java
package at.meks.calculation;

import at.meks.powerdata.SinglePowerData;

public class Battery {

    private final double batteryCapacityKwh;
    private final int lifetimeCycles;
    private double currentBatteryPower;
    private double usedKwh;

    public Battery(double batteryCapacityKwh, int lifetimeCycles) {
        this.batteryCapacityKwh = batteryCapacityKwh;
        this.lifetimeCycles = lifetimeCycles;
    }

    double consume(double kwh) {
        double removedPower = Math.min(kwh, currentBatteryPower);
        currentBatteryPower -= removedPower;
        usedKwh += removedPower;
        return removedPower;
    }

    double save(double kwh) {
        double maxSavableKwh = batteryCapacityKwh - currentBatteryPower;
        double savedKwh = Math.min(maxSavableKwh, kwh);
        currentBatteryPower = currentBatteryPower + savedKwh;
        return savedKwh;
    }

    public double usedKwh() {
        return usedKwh;
    }

    public double batteryCycles() {
        return usedKwh / batteryCapacityKwh;
    }

    public int lifetimeCycles() {
        return lifetimeCycles;
    }

    public double currentBatteryPower() {
        return currentBatteryPower;
    }
}

File: src/main/kotlin/at/meks/calculation/Battery.kt
package at.meks.calculation

import kotlin.math.min

class Battery(val batteryCapacityKwh: Double, val lifetimeCycles: Int) {

    private var currentBatteryPower: Double = 0.0
    private var usedKwh: Double = 0.0

    fun consume(kwh: Double): Double {
        val removedPower = min(kwh, currentBatteryPower)
        currentBatteryPower -= removedPower
        usedKwh += removedPower
        return removedPower
    }

    fun save(kwh: Double): Double {
        val maxSavableKwh = batteryCapacityKwh - currentBatteryPower
        val savedKwh = min(maxSavableKwh, kwh)
        currentBatteryPower += savedKwh
        return savedKwh
    }

    fun usedKwh(): Double = usedKwh

    fun batteryCycles(): Double = usedKwh / batteryCapacityKwh

    fun currentBatteryPower(): Double = currentBatteryPower

}
